### PR TITLE
Fix implicit scope implementation

### PR DESCRIPTION
**Pull request by odersky**
Change implicit scope implementation so that it can be understood and adapted to different strategies.
Based on #6824. First new commit is aa4aee6.

### CODE REVIEW

**Comment by odersky** [comment]
I don't know. I have personally no interest to change the spec here since I believe it works well. If someone wants to make a complete PR including a changed spec rules, we can consider it. But my default position will be skeptical. And, at least for Scala 3.0 we have lots of things on our plate so I am not keen on adding to this pile.

**Comment by anatoliykmetyuk** [comment]
test performance please

**Comment by milessabin** [comment]
@odersky @smarter this seems to have a semantic conflict with #6842. The change to the signature of `Symbol#isAbsent` in https://github.com/lampepfl/dotty/pull/6842/commits/a92a209760abea78d44df54aad28fad3b82aa2ed breaks its use in `Implicits.scala`.

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala [hunk 1]
```diff
@@ -541,46 +559,53 @@ trait ImplicitRunInfo { self: Run =>
...
-                  addRef(companion.select(nme.PACKAGE))
-                }
-                else if (compSym.exists)
-                  comps += companion.asSeenFrom(pre, compSym.owner).asInstanceOf[TermRef]
-              }
-              def addCompanionOf(sym: Symbol) = {
-                val companion = sym.companionModule
-                if (companion.exists) addRef(companion.termRef)
+        def addCompanion(pre: Type, companion: Symbol) =
+          if (companion.exists && !companion.isAbsent) comps += TermRef(pre, companion)
```
Comment (milessabin) [inline compiler/src/dotty/tools/dotc/typer/Implicits.scala]: Use of `isAbsent` here breaks after https://github.com/lampepfl/dotty/pull/6842.

Comment (smarter) [reply compiler/src/dotty/tools/dotc/typer/Implicits.scala]: What do you mean by break ? The signature of `isAbsent` changed, so at the very least you need to add parens to it: `isAbsent()`

Comment (milessabin) [reply compiler/src/dotty/tools/dotc/typer/Implicits.scala]: Well, yes, precisely :-)

Maybe my terminology is off here, but no git conflict, CI passes, yet will fail if merged.

**Comment by milessabin** [comment]
LGTM. Also verified that the technique proposed to fix shapeless after this change will work.

**Pull request merged**